src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -900,9 +907,21 @@ impl<'gctx> RustcTargetData<'gctx> {
         // needs access to the target config data, create a copy so that it
         // can be found. See `rebuild_unit_graph_shared` for why this is done.
         if requested_kinds.iter().any(CompileKind::is_host) {
-            let ct = CompileTarget::new(&rustc.host)?;
-            target_info.insert(ct, host_info.clone());
-            target_config.insert(ct, gctx.target_cfg_triple(&rustc.host)?);
+            target_config.insert(host_target, gctx.target_cfg_triple(&rustc.host)?);
```

Comment:
This looks correct. 

Would you mind sharing why we need the entire infra change (`Arc` and move out rustflags from targetinfo), instead of just the change here?

Reply:
This question is basically what "How it is implemented" tries to answer in the original comment on this PR. To take another stab at explaining it here:

The `CompileKind` for artifact `Unit` starts out as `CompileKind::Target(host_target)` (assuming we're not cross compiling). The `CompileKind` for host `Unit` starts out as `CompileKind::Host`
  
This change sets up `target_info` so that lookups for `CompileKind::Target(host_target)` get the correct rustflags (as well as units with `CompileKind::Host`, which need different rustflags). Which is great provided we lookup artifact `Unit`'s with their original `CompileKind`.

`rebuild_unit_graph_shared` changes artifact `Unit`s with `CompileKind::Target(host_target)` to have `CompileKind::Host`. It does this so that we can deduplicate identical otherwise identical units and only compile them once\*.

The current infrastructure attempts to look up `rustflags` via `kind` after `rebuild_unit_graph_shared`. Thus even though we have correctly set up the `target_info` for `CompileKind::Target(host_target)` we're going to look up the flags in `host_info`, which are wrong. The infrastructure change is to push this lookup forwards to before we rebuild the unit with a different `kind`. It is also necessary because we deduplicate `Unit`s based on their hash, so even if we could somehow lookup the correct rustflags later, there wouldn't necessarily be a unique set of them (in the case where host and target artifacts have different rustflags, and a shared dependency).

\* To draw your attention to the table at the top. This is why `no --target flag`, `target_applies_to_host=false` has "Without rustflags, built with 5 invocations" rather than 6. A dependency shared between the host artifacts and the target artifacts only has to be built once because it is built the exact same way for both of them.

---

However this question caused me to double check that my understanding was correct by... trying it. I believe it is, however it turns out my test is broken and will (incorrectly) pass (fail to compile) because passing an invalid rustflag will cause `TargetInfo::new(host_target)` to fail because that command calls rustc with rustflags to query file-names/sysroot/... even though we never build a target with those rustflags.

I'll figure out how to write a version of that test without the false-pass on the partial change and push it later tonight.

Reply:
Thanks. I like the idea that `Unit` holds the exact rustflags it is going to invoke. It opens a new door that we might be able to use this information to create the acutal “build plans” for external systems.

There's only one nit left and I'll merge this :)